### ISSUE TITLE
Downloads stats are off

### ISSUE THREAD

**Issue by mcollina** [nodejs/build#3697]
I took a look at https://nodedownloads.nodeland.dev/ and it seems downloads stats for April are off.

Quite a bit of data is missing, as those are the only files available for April:

nodejs.org-access.log.20240406.json
nodejs.org-access.log.20240407.json
nodejs.org-access.log.20240418.json
nodejs.org-access.log.20240419.json
nodejs.org-access.log.20240420.json
nodejs.org-access.log.20240421.json
nodejs.org-access.log.20240423.json
nodejs.org-access.log.20240424.json
nodejs.org-access.log.20240425.json
nodejs.org-access.log.20240426.json
nodejs.org-access.log.20240427.json
nodejs.org-access.log.20240428.json
nodejs.org-access.log.20240429.json

Can somebody with access retrieve the missing data?

**Comment by targos** [comment]
![CleanShot 2024-05-02 at 11 55 08@2x](https://github.com/nodejs/build/assets/2352663/294d0aac-692b-47a0-ab43-d705fd05fd05)

There's a problem with the log processor on GCP. I don't have time to investigate it.

**Comment by targos** [comment]
If someone else with access wants to have a look: https://console.cloud.google.com/run/detail/us-central1/processlogs/metrics?project=nodejs-org

**Comment by targos** [comment]
I formalized an access request: https://github.com/nodejs/build/issues/3697#issuecomment-2095994277

**Comment by richardlau** [comment]
I currently don't have time to look into this either, but for a quick glance at GCP some things that probably warrant looking at:
- We appear to be running containers (https://console.cloud.google.com/gcr/images/nodejs-org?project=nodejs-org) hosted on gcr.io
  -  I believe the Dockerfiles for those are in:
      - [`ansible/roles/metrics/files/index-generator`](https://github.com/nodejs/build/tree/main/ansible/roles/metrics/files/index-generator)
      - [`ansible/roles/metrics/files/process-cloudflare`](https://github.com/nodejs/build/tree/main/ansible/roles/metrics/files/process-cloudflare)
      - [`ansible/roles/metrics/files/summaries`](https://github.com/nodejs/build/tree/main/ansible/roles/metrics/files/summaries)
 - They are running really old versions of Node.js (14 and 15). It should be fairly simple to update the base image for the Dockerfiles, but I don't recall how these containers were built/published to the Container Registry.
 - The Container Registry is being sunset and replaced with Artifactory: https://cloud.google.com/artifact-registry/docs/transition/prepare-gcr-shutdown
  - https://github.com/nodejs/build/blob/d578923f5f389b61b19a15e74ea5313c02fad4cd/ansible/roles/metrics/README.md?plain=1#L17 -- 16GiB sounds like plenty of memory ("should be enough") but is probably worth double checking.

**Comment by mcollina** [comment]
I'm still planning to look into it, but things got deferred.

**Comment by mcollina** [comment]
Unfortunately, I did not have time to look into this, but it looks like this is a serious issue as the last day we had the data is the 2024/5/2. Maybe somebody else can step in? I don't think I would _not_ be able to look into it in the coming weeks/months.

**Comment by mcollina** [comment]
@trivikr maybe could you take a look?

**Comment by trivikr** [comment]
> @trivikr maybe could you take a look?

Sure. I created access request at https://github.com/nodejs/build/issues/3760, and shared the email address of my Google account over Slack with @targos

**Comment by trivikr** [comment]
Direct link for viewing April stats: https://storage.googleapis.com/access-logs-summaries-nodejs/?marker=nodejs.org-access.log.20240331.json

Verified that the access logs are missing for the dates, as mentioned in the description https://github.com/nodejs/build/issues/3697#issue-2275044074 above.

<details>
<summary>Screenshot</summary>

<img width="553" alt="nodejs.org-access.log.202404" src="https://github.com/nodejs/build/assets/16024985/4a0b7c84-a71e-41d8-804c-4462e0256b0d">

</details>

**Comment by trivikr** [comment]
As per the source code, it looks like the contents of the filename are not a valid JSON

https://github.com/nodejs/build/blob/437f7e389f71e9cf10c082694fb6ddfc5c655500/ansible/roles/metrics/files/process-cloudflare/process-cloudflare.js#L142-L164

Where can I read the raw logs to find out the filename being read?

**Comment by trivikr** [comment]
Example log which throws error

```console
BUCKET cloudflare-logs-nodejs
FILENAME 20240618/20240618T225603Z_20240618T225716Z_da79c9d1.log.gz
PROCESSEDFILENAME 20240618/20240618T225603Z_20240618T225716Z
INSIDE CREATE PIPELINE
String length:  21570831
Array Length:  44744
SyntaxError: Unexpected end of JSON input 
```

Direct link to file: [20240618/20240618T225603Z_20240618T225716Z_da79c9d1.log.gz](https://fff007dcec9cb8d79c764bff6008997224f964faf473202e71cbba8-apidata.googleusercontent.com/download/storage/v1/b/cloudflare-logs-nodejs/o/20240618%2F20240618T225603Z_20240618T225716Z_da79c9d1.log.gz?jk=AS1gM8PFKDuNQymvnfXVTGRvGFs8N1H2t6sGWHkrX5-l4NaNdd6i1EWdRCCepcq-cWYzDY4AWeuvsHqWXOcX9NsN0H9iW1UR08wzHwKGElkRbuqQvhI3Cvd75vNwRo3VzXdMZoN8KBVDAW7Hfo_vklb5CpiWiljlaFmuqHJ3tF1acQS3sVJ36RtB_DPJatDypKxUzbsQV4HcENQQKhq9THoDwlYllBD9PyTMNUOT585oHlj8q70kJhP6uZyMK3F1j3FlbAw9iH3EqFRKSuoytNop8hahURvyOmVR0xHm4OwM4OCtlZMZxmm_NlYOLhkhO8ZuDm_cHFH4A7Z44fjB2PIIUT6G9sF4GU2TcidXhhLJnTAxLUYu1iRzwjPgH2xIRVSVXZn8OO0PtzTVETT8BFX25oxJZzZGQk5tIYVNN2eTmhirIIHnAkC3i-TXDmqSazZBRBK7wyigjM-qKtaSujY0ZUf7ZDJZYoouJbjiVN-SnAfPHtMcDqIbQ8em7ZsYiPnzZXG5Bf35veFHGPBfz3wiH-mpsir_xpQHxjvwCYbSN0YbgaT2RhsHa6rDx_mA_PzSnxLXr0hyYVzqQinznfuzUasmkrat-TmO7eJiKqUGWgRoXhVQmo2v-UKT1ElsCCWwN9f0VVQlbmMWLKD51lAxbrwtNtT8rmmAiyQFL8WceFYS6ISLQ1HTManiKSNE8Q5AAewnRBxK_lYywnpkCXsycNL-bsNYTphvL65W3euRj0y0Ew1BLFPQLTbSI9Bh_-vXHcwOhyOIZGXIE8affhAx_WtiuxYVyiB_jU7n_9mxryuq7aKMTZhvt4myg7M2fTnfG_1iZrwiUk_cJAJE8fR9L_EnhLp_-CI5jJk_GDXQDjquz66AS-w8yzYONJCjPl_Jfcj1FtWNc8mN0vjCNueYHMQ2xzN31KPY45YgUdOb6sxAAhGdN6SsW9zdScLIgWM5TQlsSFDKJ49VknWFFGLnLAQ9jjBXX5Eq9OeRAHzT2ln2drPdFsM-3LWD14KkBnZBTzrwubiiKer0k_uAVGFdyY9R-59VgkWb52mQVmGNo1wYbpC7Ji4lFQOjaNYl4NCPTJwMlX8VSehEm2Uj434VdVzmVt2977XUpEoV_ZSkpXuZJoKszAUkYzdLuC9j887-meDMq5DcshaKAz_FF-UTooNZTnsuA6r7eHOLgs_XDyNjGLBhwYfKUGZa7yg&isca=1)

Sample code which processes this file by reading file locally from `test.json`

```js
import { readFile } from "fs/promises";

const data = await readFile("test.json", "utf8");
for (const line of data.split("\n")) {
  try {
    JSON.parse(line);
  } catch (error) {
    const { message } = error;
    console.log({ message, line });
  }
}
```

Example output
```console
{ message: 'Unexpected end of JSON input', line: '' }
```

An easy fix would be to skip reading lines which don't have any data
```js
// ...
for (const line of data.split("\n")) {
  if (line.length === 0) {
    continue;
  }
// ...
```

**Comment by trivikr** [comment]
The SyntaxError is just logged in console but the processing continues.
It may not fix the download stats, but it'll reduce the number of error occurrences in the logs.

https://github.com/nodejs/build/blob/437f7e389f71e9cf10c082694fb6ddfc5c655500/ansible/roles/metrics/files/process-cloudflare/process-cloudflare.js#L162-L168

**Comment by trivikr** [comment]
Posted another PR to log line with length if parsing fails https://github.com/nodejs/build/pull/3766

**Comment by trivikr** [comment]
I'll resume my deep dive once fixes for SyntaxError are reviewed, merged and deployed.

**Comment by trivikr** [comment]
I've created some bug reports based on logs.

However, the access logs are not created after `20240503` and Google Cloud retains logs only for 30 days.
The logs from the successful processing `nodejs.org-access.log.20240502.json` are not available for viewing.

The next steps would be to deep dive on https://github.com/nodejs/build/issues/3771, and add more logs in `collectData`
https://github.com/nodejs/build/blob/e3e4b4a7001192e2630fcf5cd84fae9df012e014/ansible/roles/metrics/files/summaries/summaries.js#L76-L87

I can take it up after getting approvals for existing open PRs.
If I can get a write/deploy permission for Google Cloud Run, it would be helpful.

**Comment by trivikr** [comment]
Updates:
* I'm investigating if increasing request timeouts is successful in producing summaries in https://github.com/nodejs/build/issues/3771
* Once it's fixed, or in parallel, an update can be posted in process_summaries to process specific dates https://github.com/nodejs/build/issues/3780

**Comment by trivikr** [comment]
> * I'm investigating if increasing request timeouts is successful in producing summaries in https://github.com/nodejs/build/issues/3771

No action needed here, although increasing the timeout is recommended in https://github.com/nodejs/build/issues/3771#issuecomment-2190393625

The summaries are being produced since [20240629](https://storage.googleapis.com/access-logs-summaries-nodejs/?marker=nodejs.org-access.log.20240628.json). I couldn't find the exact reason, the removal of strftime seems to be the closest match.

> * an update can be posted in process_summaries to process specific dates https://github.com/nodejs/build/issues/3780

A fix for this is pending merge in https://github.com/nodejs/build/pull/3801
Once merged, I'll be able to process summaries for all missing dates.

**Comment by trivikr** [comment]
On examining the source code of `summaries`, the only thing which comes out as long pole seems to be the number of files which it needs to process, which happens in parallel https://github.com/nodejs/build/blob/aa3399059a3c7095a02221c1535e8eae50eeafec/ansible/roles/metrics/files/summaries/summaries.js#L70-L72

However, I couldn't find a clear correlation between the missing summaries and number of files processed

Test code:
```js
import { Storage } from "@google-cloud/storage";

const projectId = "nodejs-org";
const storage = new Storage({ projectId });

for (let i = 1; i < 31; i++) {
  const date = `202404${i.toString().padStart(2, "0")}`;

  const filePrefix = `${date}/`;
  const [files] = await storage.bucket("processed-logs-nodejs").getFiles({ prefix: filePrefix });

  const fileName = `nodejs.org-access.log.${date}.json`;
  const [exists] = await storage.bucket("access-logs-summaries-nodejs").file(fileName).exists();

  console.log(date, files.length, exists);
}
```

Output
```console
20240401 3611 false
20240402 3757 false
20240403 3774 false
20240404 3725 false
20240405 3895 false
20240406 3671 true
20240407 3402 true
20240408 3708 false
20240409 3781 false
20240410 3776 false
20240411 4204 false
20240412 4883 false
20240413 4659 false
20240414 4694 false
20240415 4452 false
20240416 2700 false
20240417 2674 false
20240418 2654 true
20240419 2643 true
20240420 2609 true
20240421 2631 true
20240422 2689 false
20240423 2663 true
20240424 2660 true
20240425 2652 true
20240426 2640 true
20240427 2618 true
20240428 2636 true
20240429 2754 true
20240430 2664 false
```

It could happen because of the content of the files, or some other reason.
We'll get more information once we retry running summaries for those dates after https://github.com/nodejs/build/pull/3801 is merged and deployed.

**Comment by trivikr** [comment]
The summaries are created for months of April to June, with the fix in https://github.com/nodejs/build/pull/3809
The list can be viewed at https://storage.googleapis.com/access-logs-summaries-nodejs/?marker=nodejs.org-access.log.20240401.json

Test code to call summaries
```js
import { GoogleAuth } from "google-auth-library";
import { Storage } from "@google-cloud/storage";
import { Agent } from "undici";

const projectId = "nodejs-org";
const storage = new Storage({ projectId });

const cloudRunUrl = "https://produce-summaries-kdtacnjogq-uc.a.run.app";
const auth = new GoogleAuth();
const client = await auth.getIdTokenClient(cloudRunUrl);

globalThis[Symbol.for("undici.globalDispatcher.1")] = new Agent({
  headersTimeout: 30 * 60 * 1000,
});

const start = new Date("04/01/2024");
const end = new Date("06/30/2024");

for (let date = start; date <= end; date.setDate(date.getDate() + 1)) {
  const dateString = date.toISOString().split("T")[0].replace(/-/g, "");

  const fileName = `nodejs.org-access.log.${dateString}.json`;
  const [exists] = await storage.bucket("access-logs-summaries-nodejs").file(fileName).exists();

  if (!exists) {
    const startTime = Date.now();
    console.log(`Fetching for ${dateString}`);
    // Make a call to the API to populate the file
    const url = `${cloudRunUrl}/date/${dateString}`;

    try {
      const token = await client.idTokenProvider.fetchIdToken(cloudRunUrl);
      const response = await fetch(url, {
        method: "POST",
        headers: { Authorization: `Bearer ${token}` },
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      console.log(`Successful in ${Date.now() - startTime}ms for ${dateString}`);
    } catch (error) {
      console.error(`Error in ${Date.now() - startTime}ms for ${dateString}`, error);
    }
  }
}
```

The script was run in parallel for different dates. This issue can be closed.

**Comment by trivikr** [comment]
After the logs are replaced, the https://nodedownloads.nodeland.dev/ is showing a bump from May though.

<details>
<summary>Screenshot</summary>

<img width="557" alt="Screenshot 2024-07-06 at 12 30 25 PM" src="https://github.com/nodejs/build/assets/16024985/209eeb16-4e9c-4952-8659-68be925c2eb4">

</details>

A new issue can be created with details, if there's an issue with the replaced metric summaries.
If not, the issue needs to be reported on https://github.com/mcollina/nodejs-download-stats/

**Comment by mcollina** [comment]
thanks @trivikr!!!
